src/Development/IDE/Core/Rules.hs

```diff
@@ -112,9 +112,59 @@ getDefinition :: NormalizedFilePath -> Position -> Action (Maybe Location)
 getDefinition file pos = fmap join $ runMaybeT $ do
     opts <- lift getIdeOptions
     spans <- useE GetSpanInfo file
-    pkgState <- hscEnv <$> useE GhcSession file
-    let getHieFile x = useNoFile (GetHieFile x)
-    lift $ AtPoint.gotoDefinition getHieFile opts pkgState (spansExprs spans) pos
+    lift $ AtPoint.gotoDefinition (getHieFile file) opts (spansExprs spans) pos
+
+getHieFile
+  :: NormalizedFilePath -- ^ file we're editing
+  -> Module -- ^ module dep we want info for
+  -> Action (Maybe (HieFile, FilePath)) -- ^ hie stuff for the module
+getHieFile file mod = do
+  TransitiveDependencies {transitiveNamedModuleDeps} <- use_ GetDependencies file
+  case find (\x -> nmdModuleName x == moduleName mod) transitiveNamedModuleDeps of
+    Just NamedModuleDep{nmdFilePath=nfp} -> do
+        let modPath = fromNormalizedFilePath nfp
+        (_diags, hieFile) <- getHomeHieFile nfp
+        return $ (, modPath) <$> hieFile
+    _ -> getPackageHieFile mod file
+
+
+getHomeHieFile :: NormalizedFilePath -> Action ([a], Maybe HieFile)
+getHomeHieFile f = do
+  pm <- use_ GetParsedModule f
+  let normal_hie_f = toNormalizedFilePath hie_f
+      hie_f = ml_hie_file $ ms_location $ pm_mod_summary pm
+  mbHieTimestamp <- use GetModificationTime normal_hie_f
+  srcTimestamp   <- use_ GetModificationTime f
+  case (mbHieTimestamp, srcTimestamp) of
+    (Just hie, src)
+      | comparing modificationTime hie src == GT -> pure ()
+    _ -> do
+      -- typecheck generates a .hie file as a side effect
+      void $ use_ TypeCheck f
+  hf <- liftIO $ loadHieFile hie_f
+  return ([], Just hf)
+
+getPackageHieFile :: Module             -- ^ Package Module to load .hie file for
+                  -> NormalizedFilePath -- ^ Path of home module importing the package module
+                  -> Action (Maybe (HieFile, FilePath))
+getPackageHieFile mod file = do
+    pkgState  <- hscEnv <$> use_ GhcSession file
+    IdeOptions {..} <- getIdeOptions
+    let unitId = moduleUnitId mod
+    case lookupPackageConfig unitId pkgState of
+        Just pkgConfig -> do
+            hieFile <- liftIO $ optLocateHieFile optPkgLocationOpts pkgConfig mod
+            path    <- liftIO $ optLocateSrcFile optPkgLocationOpts pkgConfig mod
+            case (hieFile, path) of
+                (Just hiePath, Just modPath) -> liftIO $ do
+                    -- deliberately loaded outside the Shake graph
+                    -- to avoid dependencies on non-workspace files
+                    foundIt <- IO.doesFileExist hiePath
```

Comment:
The intention of `optLocateHieFile` was to return `Nothing` if the file does not exist, at least that’s how we use it in DAML.

Reply:
Right, I'll fix and add a comment to make it explicit